### DISCUSSION TITLE
Generic Controls

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#13480] [Ideas]
# Generic Controls

This is an idea I've been considering for a while. When C# first came about there were no generics. WPF was mostly implemented without them as well. This is a problem for some controls like the NumericUpDown/NumberBox or even DatePicker.

.NET now has many, many types. These should be used under slightly different scenarios:

 1. NumberBox : Should use either `double` or `decimal` but also could be `int`, `long`, etc. We've currently settled on `decimal` as that supports all cases exactly but still requires type conversion in most cases. Not ideal.
 2. Calendar : .NET has `DateTime`, `DateTimeOffset` and now `DateOnly`. Again, we have to decide what the best type is here and unfortunatley for legacy reasons its DateTime. That isn't great and it now should really be DateOnly but apps may need DateTimeOffset as well.

.NET is now far enough along I think these problems can be fixed by introducing a new concept called "Generic Controls".

## NumberBox

Note: This assumes a port from WinUI3 for this control first. That is something I mean to do eventually along with making the number parsing pipeline customizable which wasn't supported in that control originally.

With .NET 7 we now have the INumber<T> interface. This combined with a new concept for generic controls can solve this problem in the case of a NumberBox.

```csharp
public class NumberBox<T> where T : struct, INumber<T>, IParsable<T>
{
}

public class NumberBox : NumberBox<decimal>
{
}
```

This would allow the control to support ANY numeric type. In addition, you have a version you can use with no type parameter that will default to decimal. This seems like the best of both worlds and can be used anywhere.

## Calendar

Calendar-related controls can't get this functionality yet because there is no IDate interface in .NET yet. My bet is that is something that would be accepted upstream though. It is also very easy to do.

Upstream discussion: https://github.com/dotnet/runtime/issues/97622

## Feedback

What are everyone's thoughts to this idea?

**Comment by robloo** [comment]
Making NumericUpDown a Generic Control supporting decimal, double, float, etc. Should be possible in the Avalonia 12 timing. It requires .NET 7+ and I believe the plan is .NET 8 minimum or so?

Keeping an eye on this.

**Comment by gehongyan** [comment]
It sounds like some attempts Ursa has made on `NumericUpDownBase<T>`.

https://github.com/irihitech/Ursa.Avalonia/blob/a5e1d6efa4cfac38e46d56459088540ab980a986/src/Ursa/Controls/NumericUpDown/NumericUpDownBase.cs#L374C23-L821
